File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.py
class Solution:
    def findLexSmallestString(self, s: str, a: int, b: int) -> str:
        def addA(s, a):
            new_s = ''
            for i,c in enumerate(s):
                if i & 1:
                    new_s += str((int(c) + a) % 10)
                else:
                    new_s += c
            return new_s
        
        
        def rotate(s, b):
            return s[-b:] + s[:-b]
        
        
        dq, seen, smallest = deque([s]), set(), s
        while dq:
            curr = dq.popleft()
            if curr < smallest:
                smallest = curr
            if curr not in seen:
                seen.add(curr)
                dq.append(addA(curr, a))
                dq.append(rotate(curr, b))
        return smallest

File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.js
/**
 * @param {string} s
 * @param {number} a
 * @param {number} b
 * @return {string}
 */
var findLexSmallestString = function(s, a, b) {
  const queue = [s];
  const set = new Set();
  let smallest = s;
  
  while (queue.length) {
    const curr = queue.shift();
    if (curr < smallest) smallest = curr;
    if (!set.has(curr)) {
      set.add(curr);
      queue.push(addA(curr, a), rotate(curr, b));
    }
  }
  
  return smallest;
};

const addA = (s, a) => {
  let newS = '';
  
  for (let i = 0; i < s.length; ++i) {
    newS += (i & 1) ? (+s[i] + a) % 10 : s[i];
  }
  
  return newS;
}

const rotate = (s, b) => s.slice(-b) + s.slice(0, -b);